streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -495,6 +501,41 @@ private boolean isLastIndex(final int index) {
             return unpackedReversedTimestampAndValueSizes.get(index).timestamp == minTimestamp;
         }
 
+        private void truncateRecordsToTimestamp(final long timestamp) {
+            if (timestamp <= minTimestamp) {
+                // delete everything in this current segment by replacing it with a degenerate segment
+                initializeWithRecord(new ValueAndValueSize(null), timestamp, timestamp);
+                return;
+            }
+
+            final SegmentSearchResult searchResult = find(timestamp, false);
+            // all records with later timestamps should be removed
+            int fullRecordsToTruncate = searchResult.index();
+            // additionally remove the current record as well, if its validFrom equals the
+            // timestamp to truncate to
+            if (searchResult.validFrom() == timestamp) {
+                fullRecordsToTruncate++;
+            }
+
+            if (fullRecordsToTruncate == 0) {
+                // no records to remove; update nextTimestamp and return
+                nextTimestamp = timestamp;
+                ByteBuffer.wrap(segmentValue, 0, TIMESTAMP_SIZE).putLong(0, timestamp);
+                return;
+            }
+
+            final int valuesLengthToRemove = cumulativeValueSizes.get(fullRecordsToTruncate - 1);
+            final int timestampAndValueSizesLengthToRemove = (TIMESTAMP_SIZE + VALUE_SIZE) * fullRecordsToTruncate;
+            final int newSegmentLength = segmentValue.length - valuesLengthToRemove - timestampAndValueSizesLengthToRemove;
+            segmentValue = ByteBuffer.allocate(newSegmentLength)
+                .putLong(timestamp) // update nextTimestamp as part of truncation
```

Comment:
I think this is "strictly" not correct (we lost the original `validTo` / `next_timestamp` what was overwritten by moving a record into this segment in the partial write); but it also does not matter, because we actually have a put() at hand that will update `next_timestamp` right after we finished the cleanup again, right? (not sure if we should document this with a comment?)

Reply:
> but it also does not matter, because we actually have a put() at hand that will update next_timestamp right after we finished the cleanup again

Yeah that's correct. My intention with updating `nextTimestamp` here even though it will be immediately updated again once we return from this method (back to `insertAsLatest()`) is so that the state of the segment is always "consistent," which makes the code easier to reason about IMO. Another way to think about it is that when an inconsistency is detected from `insertAsLatest()`, we respond by truncating all contents in this segment up to `timestamp` (the timestamp of the record which is being inserted). This truncation involves updating `nextTimestamp` for the segment to be `timestamp`.

Do you have thoughts on what additional comments (and where) would help clarify this? I had hoped that the method name `truncateRecordsToTimestamp` and this comment stating that we are updating `nextTimestamp` to `timestamp` would be sufficient, but it appears not :)

Reply:
Yeah -- not 100% sure -- guess my idea was just to avoid potential confusion if somebody read this code in the future (but also always tricky to not confuse readers with too much detail). Maybe something like:
```
// update nextTimestamp as part of truncation -- this a surrogate value that just fixed inconsistency without being actually correct; but we have a pending put() in place that will set the correct value later 
```

Just an idea.

Reply:
I opted to add a javadoc to the method -- LMK whether you think it clears up the confusion here. 

I think it might be a bit too granular to add the explanation into this line itself because there's also another place in the method (the first if-condition were we replace the entire segment with a degenerate one if the truncation timestamp is less than or equal to the minTimestamp of the segment) where readers are likely to have the same question if it's not already clear that nextTimestamp will be updated to the truncation timestamp.